### PR TITLE
Logging/388 reference data overrides validation process

### LINKED ISSUE #388: Reference data overrides validation process
At startup, Conseil should parse the metadata overrides and highlight the following things in the log:

- [ ] warning on table being present in the SQL schema, but not in the metadata overrides structure.
- [ ] warning on field not being explicitly mentioned with a minimum of a `visible` attribute while being present in the SQL schema.
- [ ] error on an object being listed in the metadata overrides file, but not appearing in the SQL schema.

### PR DESCRIPTION
**Pull request by dorians**
closes #388
* metadata are precached on Conseil startup
* log if there is no overwritten metadata or there is an override which overrides nothing
* added check if network or platform are not hidden in places where that check was missing
* rewrite tests (avoid stubbing in exchange for real implementation)

### CODE REVIEW

File: src/main/scala/tech/cryptonomic/conseil/generic/chain/PlatformDiscoveryOperations.scala [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package tech.cryptonomic.conseil.generic.chain
+
+import tech.cryptonomic.conseil.generic.chain.DataTypes.AttributesValidationError
+import tech.cryptonomic.conseil.generic.chain.PlatformDiscoveryTypes.{Attribute, AttributeCacheConfiguration, Entity}
+
+import scala.concurrent.Future
+
```
Comment (piotrkosecki) [inline src/main/scala/tech/cryptonomic/conseil/generic/chain/PlatformDiscoveryOperations.scala]: please add javadoc comments

File: src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala [hunk 2]
```diff
@@ -10,24 +10,29 @@ sealed trait Path {
...
 // case class representing a path for a platform
 case class PlatformPath(platform: String, up: EmptyPath = EmptyPath()) extends Path {
   override def addLevel(networkPath: String): NetworkPath = NetworkPath(networkPath, this)
+  override def toString: String = platform
 }
 
 // case class representing a path for a network
 case class NetworkPath(network: String, up: PlatformPath) extends Path {
   override def addLevel(entity: String): EntityPath = EntityPath(entity, this)
+  override def toString: String = s"${up.toString} / $network"
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala]: No need to call `up.toString` explicitly, it's done by the string interpolator

File: src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala [hunk 3]
```diff
@@ -10,24 +10,29 @@ sealed trait Path {
...
 }
 
 // case class representing a path for an entity
 case class EntityPath(entity: String, up: NetworkPath) extends Path {
   override def addLevel(attribute: String): AttributePath = AttributePath(attribute, this)
+  override def toString: String = s"${up.toString} / $entity"
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala]: Same here

File: src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala [hunk 4]
```diff
@@ -10,24 +10,29 @@ sealed trait Path {
...
 }
 
 // case class representing a path for an attribute
 case class AttributePath(attribute: String, up: EntityPath) extends Path {
   override def addLevel(nextLabel: String): Path = throw new NotImplementedError()
+  override def toString: String = s"${up.toString} / $attribute"
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala]: and here

File: src/main/scala/tech/cryptonomic/conseil/metadata/UnitTransformation.scala [hunk 5]
```diff
@@ -69,4 +94,14 @@ class UnitTransformation(overrides: MetadataConfiguration) {
         currencySymbolCode = overrideAttribute.flatMap(_.currencySymbolCode)
       )
     }
+
+  private def logDifferences(paths: List[Path], overriddenPaths: List[Path]): Unit = {
+    paths
+      .filterNot(overriddenPaths.toSet)
+      .foreach(logger.warn("""There're missing metadata overrides for "{}"""", _))
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/metadata/UnitTransformation.scala]: Personal taste:
I'd rather make a single string list with `.mkString` and then issue a single `log.warn` with that, both here and immediately after

Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/metadata/UnitTransformation.scala]: Is having paths that are not overridden considered something to get worried about?
If that's not the case I'd suggest using `logger.info` and write the message in softer tones (e.g. "these paths were not overridden: ...")

File: src/main/scala/tech/cryptonomic/conseil/metadata/UnitTransformation.scala [hunk 6]
```diff
@@ -69,4 +94,14 @@ class UnitTransformation(overrides: MetadataConfiguration) {
         currencySymbolCode = overrideAttribute.flatMap(_.currencySymbolCode)
       )
     }
+
+  private def logDifferences(paths: List[Path], overriddenPaths: List[Path]): Unit = {
+    paths
+      .filterNot(overriddenPaths.toSet)
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/metadata/UnitTransformation.scala]: Another way to formulate is 
```scala
paths.toSet diff overriddenPaths.toSet
```

File: src/main/scala/tech/cryptonomic/conseil/routes/Data.scala [hunk 7]
```diff
@@ -35,27 +35,26 @@ class Data(config: PlatformsConfiguration, queryProtocolPlatform: DataPlatform,
...
           validationResult.map { validQuery =>
-            platformNetworkValidation(platform, network) {
-              queryProtocolPlatform.queryWithPredicates(platform, entity, validQuery).map { queryResponseOpt =>
-                queryResponseOpt.map { queryResponses =>
-                  QueryResponseWithOutput(
-                    queryResponses,
-                    validQuery.output
-                  )
-                }
+            queryProtocolPlatform.queryWithPredicates(platform, entity, validQuery).map { queryResponseOpt =>
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/routes/Data.scala]: I can't see platformNetworkValidation being done anymore? Is that intended?

File: src/test/scala/tech/cryptonomic/conseil/metadata/TestPlatformDiscoveryOperations.scala [hunk 8]
```diff
@@ -0,0 +1,35 @@
+package tech.cryptonomic.conseil.metadata
+
+import tech.cryptonomic.conseil.generic.chain.{DataTypes, PlatformDiscoveryOperations, PlatformDiscoveryTypes}
+import tech.cryptonomic.conseil.generic.chain.PlatformDiscoveryTypes.{Attribute, Entity}
+
+import scala.concurrent.Future
+
+class TestPlatformDiscoveryOperations extends PlatformDiscoveryOperations {
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/metadata/TestPlatformDiscoveryOperations.scala]: What about some very basic scaladoc here?

File: src/test/scala/tech/cryptonomic/conseil/metadata/TransparentUnitTransformation.scala [hunk 9]
```diff
@@ -0,0 +1,26 @@
+package tech.cryptonomic.conseil.metadata
+
+import tech.cryptonomic.conseil.config.MetadataConfiguration
+import tech.cryptonomic.conseil.generic.chain.PlatformDiscoveryTypes
+
+class TransparentUnitTransformation(overrides: MetadataConfiguration = MetadataConfiguration(Map.empty))
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/metadata/TransparentUnitTransformation.scala]: What about some very basic scaladoc here?

Comment (ivanopagano) [reply src/test/scala/tech/cryptonomic/conseil/metadata/TransparentUnitTransformation.scala]: Also, I guess we can use an object here

```scala
object TransparentUnitTransform extends UnitTransform(MetadataConfiguration(Map.empty) {
  ...
}
```

File: src/main/scala/tech/cryptonomic/conseil/routes/Data.scala (hunk 7 above)
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/routes/Data.scala]: It's done a few lines above as `pathValidation(path)`.

File: src/main/scala/tech/cryptonomic/conseil/metadata/UnitTransformation.scala (hunk 5 above)
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/metadata/UnitTransformation.scala]: 1. I would prefer to have a single log too but it would require some fancy collecting so I decided to do it that way in order to keep code simple. 
2. According to the task description, it's something we need to worry about. @anonymoussprocket requested explicitly that it should be `warn` and `error`.

File: src/test/scala/tech/cryptonomic/conseil/metadata/MetadataServiceTest.scala [hunk 10]
```diff
@@ -4,39 +4,50 @@ import akka.http.scaladsl.testkit.ScalatestRouteTest
...
         )
+      ),
+      new UnitTransformation(MetadataConfiguration(metadataOverridesConfiguration)),
+      cacheOverrides,
+      platformDiscoveryOperations
+    )
+
+  override protected def beforeEach(): Unit = {
+    super.beforeEach()
+    platformDiscoveryOperations.clear()
```
Comment (piotrkosecki) [inline src/test/scala/tech/cryptonomic/conseil/metadata/MetadataServiceTest.scala]: instead of that you could just mix in `OneInstancePerTest`, but I guess it depends on preferences

File: src/test/scala/tech/cryptonomic/conseil/metadata/TestPlatformDiscoveryOperations.scala [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+  var entities: List[Entity] = List.empty
+  var attributes: List[Attribute] = List.empty
+
+  def addEntity(entity: Entity): Unit = entities = entities :+ entity
```
Comment (piotrkosecki) [inline src/test/scala/tech/cryptonomic/conseil/metadata/TestPlatformDiscoveryOperations.scala]: maybe add `this.syncronized{}` to those operations?

File: src/test/scala/tech/cryptonomic/conseil/tezos/DataTypesTest.scala [hunk 12]
```diff
@@ -4,20 +4,41 @@ import java.sql.Timestamp
...
+      new TransparentUnitTransformation(),
+      cacheOverrides,
+      platformDiscoveryOperations
+    )
+  }
+
+  val testEntityPath = EntityPath("testEntity", NetworkPath("alphanet", PlatformPath("tezos")))
   val testEntity = Entity("testEntity", "Test Entity", 0)
 
+  override protected def beforeEach(): Unit = platformDiscoveryOperations.clear()
```
Comment (piotrkosecki) [inline src/test/scala/tech/cryptonomic/conseil/tezos/DataTypesTest.scala]: as mentioned above this can be exchanged with mixing in trait `OneInstancePerTest`, also why in this case there is no `super.beforeEach?`

Comment (dorians) [reply src/test/scala/tech/cryptonomic/conseil/tezos/DataTypesTest.scala]: done

File: src/test/scala/tech/cryptonomic/conseil/metadata/TestPlatformDiscoveryOperations.scala (hunk 11 above)
Comment (dorians) [reply src/test/scala/tech/cryptonomic/conseil/metadata/TestPlatformDiscoveryOperations.scala]: done

File: src/test/scala/tech/cryptonomic/conseil/metadata/MetadataServiceTest.scala (hunk 10 above)
Comment (dorians) [reply src/test/scala/tech/cryptonomic/conseil/metadata/MetadataServiceTest.scala]: done

File: src/test/scala/tech/cryptonomic/conseil/metadata/TransparentUnitTransformation.scala (hunk 9 above)
Comment (dorians) [reply src/test/scala/tech/cryptonomic/conseil/metadata/TransparentUnitTransformation.scala]: done

File: src/test/scala/tech/cryptonomic/conseil/metadata/TestPlatformDiscoveryOperations.scala (hunk 8 above)
Comment (dorians) [reply src/test/scala/tech/cryptonomic/conseil/metadata/TestPlatformDiscoveryOperations.scala]: done

File: src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala (hunk 4 above)
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala]: and done

File: src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala (hunk 3 above)
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala]: done

File: src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala (hunk 2 above)
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/metadata/Path.scala]: done

File: src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala [hunk 13]
```diff
@@ -1,48 +1,75 @@
...
+      Await
+        .result(
+          platformDiscoveryOperations
+            .getTableAttributes(path.entity)
+            .map(
+              _.map(attributes => transformation.overrideAttributes(path, attributes))
+            ),
+          1 second
+        )
+        .map(path -> _)
```
[the remark this answers is not in the document]
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala]: done

File: src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala [hunk 14]
```diff
@@ -1,48 +1,75 @@
...
+  }
+
+  private val attributes: Map[EntityPath, List[Attribute]] = {
+    val networkPaths = entities.flatMap {
+      case (networkPath: NetworkPath, entities: List[Entity]) =>
+        entities.map(entity => networkPath.addLevel(entity.name))
+    }.toSet
+
+    networkPaths.flatMap { path =>
+      Await
```
[the remark this answers is not in the document]
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala]: done

File: src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala [hunk 15]
```diff
@@ -1,48 +1,75 @@
...
+
+  private val networks = platforms.map { platform =>
+    platform.path -> transformation.overrideNetworks(platform.path, ConfigUtil.getNetworks(config, platform.name))
+  }.toMap
+
+  private val entities = {
+    val allEntities = Await.result(platformDiscoveryOperations.getEntities, 1 second)
+
+    networks.values.flatten
+      .map(_.path)
```
[the remark this answers is not in the document]
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala]: done

File: src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala [hunk 16]
```diff
@@ -1,48 +1,75 @@
...
-  def isAttributeValid(entity: String, attribute: String): Future[Boolean] =
-    tezosPlatformDiscoveryOperations.isAttributeValid(entity, attribute)
+  private val platforms = transformation.overridePlatforms(ConfigUtil.getPlatforms(config))
+
+  private val networks = platforms.map { platform =>
+    platform.path -> transformation.overrideNetworks(platform.path, ConfigUtil.getNetworks(config, platform.name))
+  }.toMap
+
+  private val entities = {
+    val allEntities = Await.result(platformDiscoveryOperations.getEntities, 1 second)
```
[the remark this answers is not in the document]
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala]: That timeout is for a single call to the database. `1 sec` seems to be pretty long. I'm afraid of setting that value too high because when the database needs over 1 sec to answer, it usually means that something horribly happens and prefilling the cache won't succeed anyway so I try to fail fast. Anyway, we don't have metrics which could prove or disprove any point so I set it to 2 s. and we'll see what'll happen :)

[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala]: I was actually more interested in a solution that would handle the error and pass it back to the user with a clear information on what went wrong, but that might be better handled later?

Personally I think that the database might be unavailable for more than 1 sec with no critical situation going on, any network hiccup could bring unexpected consequences, or maybe a slightly overloaded system busy starting multiple services?
To be honest I didn't look at time tables to figure out what might be reasonable, and the only good answer would be to observe the system and measure any potential incident to extract real values.
tl;dr; I would've chosen something less strict but it's not really better than what you did, so let's say it's fine.

File: src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/metadata/MetadataService.scala]: 🤔
